devices/src/pvmemcontrol.rs

```diff
@@ -0,0 +1,819 @@
+// Copyright © 2024 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use num_enum::TryFromPrimitive;
+use pci::{
+    BarReprogrammingParams, PciBarConfiguration, PciBarPrefetchable, PciBarRegionType,
+    PciClassCode, PciConfiguration, PciDevice, PciDeviceError, PciHeaderType, PciSubclass,
+};
+use std::{
+    collections::HashMap,
+    ffi::CString,
+    io, result,
+    sync::{Arc, Barrier, Mutex, RwLock},
+};
+use thiserror::Error;
+use vm_allocator::{page_size::get_page_size, AddressAllocator, SystemAllocator};
+use vm_device::{BusDeviceSync, Resource};
+use vm_memory::{
+    bitmap::AtomicBitmap, Address, ByteValued, Bytes, GuestAddress, GuestAddressSpace, GuestMemory,
+    GuestMemoryAtomic, GuestMemoryError, GuestMemoryMmap, Le32, Le64,
+};
+use vm_migration::{Migratable, MigratableError, Pausable, Snapshot, Snapshottable, Transportable};
+
+const PVMEMCONTROL_VENDOR_ID: u16 = 0x1ae0;
+const PVMEMCONTROL_DEVICE_ID: u16 = 0x0087;
+
+const PVMEMCONTROL_SUBSYSTEM_VENDOR_ID: u16 = 0x1ae0;
+const PVMEMCONTROL_SUBSYSTEM_ID: u16 = 0x011F;
+
+const MAJOR_VERSION: u64 = 1;
+const MINOR_VERSION: u64 = 0;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    // device errors
+    #[error("Guest gave us bad memory addresses: {0}")]
+    GuestMemory(#[source] GuestMemoryError),
+    #[error("Guest sent us invalid request")]
+    InvalidRequest,
+
+    #[error("Guest sent us invalid command")]
```

Comment:
You can show the invalid command in the error message to help debugging. The same goes for other errors.

Reply:
Fixed